### PR TITLE
[SYCL-MLIR] Add sycl::vec support in the SYCL dialect

### PR DESCRIPTION
**Pull request by victor-eds**
Add parsing, printing and lowering support for `sycl::vec`.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

**Comment by victor-eds** [comment]
First of a PR series on supporting `sycl::vec`.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOpsTypes.h [hunk 1]
```diff
@@ -299,6 +299,34 @@ struct GroupTypeStorage : public TypeStorage {
...
+
+  static KeyTy getKey(const KeyTy &Key) { return KeyTy{Key}; }
+
+  static VecTypeStorage *construct(TypeStorageAllocator &Allocator,
+                                   const KeyTy &Key) {
+    return new (Allocator.allocate<VecTypeStorage>()) VecTypeStorage(Key);
+  }
+
+  mlir::Type DataT;
+  int NumElements;
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOpsTypes.h:326]: Chose `int` as this is the type used in the template.

**Review by etiotto — APPROVED** [review]

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir [hunk 2]
```diff
@@ -70,4 +70,7 @@ module {
   func.func @test_nd_item(%arg0: !sycl.nd_item<[1], (!sycl.item<[1, true], (!sycl.item_base<[1, true], (!sycl.range<1>, !sycl.id<1>, !sycl.id<1>)>)>, !sycl.item<[1, false], (!sycl.item_base<[1, false], (!sycl.range<1>, !sycl.id<1>)>)>, !sycl.group<[1], (!sycl.range<1>, !sycl.range<1>, !sycl.range<1>, !sycl.id<1>)>)>) {
     return
   }
+  func.func @test_vec(%arg0: !sycl.vec<[f32, 4], (vector<4xf32>)>) {
```
Comment (whitneywhtsang) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:73]: Missing check line.

File: mlir-sycl/test/Dialect/IR/SYCL/types.mlir [hunk 3]
```diff
@@ -145,3 +145,19 @@ func.func @_Z7group_1N2cl4sycl5groupILi1EEE(%arg0: !sycl_group_1_) attributes {l
 func.func @_Z7group_2N2cl4sycl5groupILi2EEE(%arg0: !sycl_group_2_) attributes {llvm.linkage = #llvm.linkage<external>} {
   return
 }
+
```
Comment (whitneywhtsang) [inline mlir-sycl/test/Dialect/IR/SYCL/types.mlir:148]:
```suggestion

// -----

```

**Review by whitneywhtsang — CHANGES_REQUESTED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**